### PR TITLE
AKCORE-72: Refactored share partition manager to handle locks better

### PR DESCRIPTION
**Pull request by apoorvmittal10**
- Fixed lock handling and release for share partition (finally is incorrect, share partition lock pre-emptive release is incorrect)
- Refactored code.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 1]
```diff
@@ -1262,6 +1262,8 @@ class KafkaApis(val requestChannel: RequestChannel,
         clientMetadata
       )
 
+      // TODO: We cna simplify the code by removing the need for the additional max bytes map. Rather
```
Comment (AndrewJSchofield) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1265]: cna? can!

**Review by AndrewJSchofield — APPROVED** [review]
I would like to see a test that demonstrates the change was valuable.

**Comment by apoorvmittal10** [comment]
> I would like to see a test that demonstrates the change was valuable.

Though @adixitconfluent is working on adding the tests but as the code is only processed by an active check lock in the queue hence we should not see the benefits right away. But as we move towards reading data parallelly from  the queue with the lock just on share partition then we should see the benefits. Do you think it's sensible to merge the PR for now?

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 1 above)
Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1265]: I will do that, thanks for pointing.

File: core/src/main/java/kafka/server/SharePartitionManager.java [hunk 2]
```diff
@@ -99,101 +88,122 @@ public CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionD
...
+                        partitionMaxBytes,
+                        Optional.empty()));
+                }
+            });
+
+            if (topicPartitionData.isEmpty()) {
+                // No locks for share partitions could be acquired, so we complete the request and
+                // will re-fetch for the client in next poll.
+                shareFetchPartitionData.future.complete(Collections.emptyMap());
+                return;
```
Comment (adixitconfluent) [inline core/src/main/java/kafka/server/SharePartitionManager.java:137]: shouldn't we also call  `releaseProcessFetchQueueLock()` method to release fetch queue lock here?

**Review by adixitconfluent — CHANGES_REQUESTED** [review]

File: core/src/main/java/kafka/server/SharePartitionManager.java (hunk 2 above)
Comment (apoorvmittal10) [reply core/src/main/java/kafka/server/SharePartitionManager.java:137]: I didn't add the check as anyways some request should be in-flight. But probably shall do that, we might need to do better here later.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 1 above)
Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Done.

File: core/src/main/java/kafka/server/SharePartitionManager.java (hunk 2 above)
Comment (apoorvmittal10) [reply core/src/main/java/kafka/server/SharePartitionManager.java:142]: Added.

**Review by adixitconfluent — APPROVED** [review]
LGTM

**Pull request merged**